### PR TITLE
Upload pipeline logs in CI pipeline

### PR DESCRIPTION
**Pull request by Allda**
CI pipeline uploads pipeline logs using a upload script + Pyxis API. The
upload step is defined in finally block so it is always executer no
matter if previous steps passed or failed.

JIRA: ISV-927

### CODE REVIEW

File: ansible/roles/operator-pipeline/templates/openshift/pipelines/operator-ci-pipeline.yml [hunk 1]
```diff
@@ -290,3 +290,21 @@ spec:
         - name: source
           workspace: pipeline
           subPath: src
+
+  finally:
+    - name: upload-pipeline-logs
+      taskRef:
+        name: upload-pipeline-logs
```
Comment (amisstea) [inline ansible/roles/operator-pipeline/templates/openshift/pipelines/operator-ci-pipeline.yml:297]: Where's the definition of this task?

Comment (Allda) [reply ansible/roles/operator-pipeline/templates/openshift/pipelines/operator-ci-pipeline.yml:297]: I forgot to add this file. Added..

File: ansible/roles/operator-pipeline/templates/openshift/tasks/upload-artifact.yml [hunk 2]
```diff
@@ -37,35 +37,92 @@ spec:
...
+        upload-artifacts \
+          --cert-project-id "$(params.cert_project_id)" \
+          --operator-version "$(params.bundle_version)" \
+          --operator-package-name "$(params.package_name)" \
+          --certification-hash "$(params.md5sum)" \
+          --pyxis-url "$(params.pyxis_url)" \
+          --path "$(params.artifacts_dir)" \
+          --type preflight-artifacts \
+          --output preflight-artifacts.json \
+          --verbose
```
Comment (wcheang) [inline ansible/roles/operator-pipeline/templates/openshift/tasks/upload-artifact.yml:128]: No need to output/return the artifacts id/url?

Comment (Allda) [reply ansible/roles/operator-pipeline/templates/openshift/tasks/upload-artifact.yml:128]: No, there can be many artifacts produced by the preflight.

**Comment by Allda** [comment]
@amisstea I split a task into 2 steps and the first one uses the official tekton image as you suggested. Can you please check it again?

File: ansible/roles/operator-pipeline/templates/openshift/tasks/upload-artifact.yml [hunk 3]
```diff
@@ -37,35 +37,92 @@ spec:
...
+          valueFrom:
+            secretKeyRef:
+              name: $(params.pyxis_api_key_secret)
+              key: PYXIS_API_KEY
+      script: |
+        #! /usr/bin/env bash
+        PREFLIGHT_RESULTS_EXISTS="$(params.preflight_results_exists)"
+        if [ $PREFLIGHT_RESULTS_EXISTS == "true" ]; then
+          echo "Preflight results already exists - skipping"
+          echo "none" | tee $(results.result_url.path)
```
Comment (amisstea) [inline ansible/roles/operator-pipeline/templates/openshift/tasks/upload-artifact.yml:78]: Should we provide the url to the existing results instead? If not then this special value should be documented in the description of the result.

This applies to the `log_url` result as well.

File: ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      default: pyxis-api-secret
+    - name: pyxis_url
+      default: https://catalog.redhat.com/api/containers/
+    - name: md5sum
+    - name: cert_project_id
+    - name: bundle_version
+    - name: package_name
+    - name: pipeline_name
+  results:
+    - name: pipeline_log_url
```
Comment (amisstea) [inline ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml:18]: Please add descriptions for all params and results.

File: ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  steps:
+    - name: get-pipeline-logs
+      image: registry.redhat.io/openshift-pipelines/pipelines-cli-tkn-rhel8:0.19.0-2
+      script: |
+        #! /usr/bin/env bash
+        set -x
```
Comment (amisstea) [inline ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml:24]: Do we need the `-e` switch as well to ensure the script errors out for non-zero exit codes?

File: ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      env:
+        - name: PYXIS_API_KEY
+          valueFrom:
+            secretKeyRef:
+              name: $(params.pyxis_api_key_secret)
+              key: PYXIS_API_KEY
+      script: |
+        #! /usr/bin/env bash
+        echo "Uploading pipeline logs"
+        set -x
```
Comment (amisstea) [inline ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml:41]: Same thing here about `-e`.

**Review by amisstea — CHANGES_REQUESTED** [review]

File: ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml (hunk 4 above)
Comment (Allda) [reply ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml]: Added

File: ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml (hunk 5 above)
Comment (Allda) [reply ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml]: added

File: ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml (hunk 6 above)
Comment (Allda) [reply ansible/roles/operator-pipeline/templates/openshift/tasks/upload-pipeline-logs.yml]: added

File: ansible/roles/operator-pipeline/templates/openshift/tasks/upload-artifact.yml (hunk 3 above)
Comment (Allda) [reply ansible/roles/operator-pipeline/templates/openshift/tasks/upload-artifact.yml:78]: Oh.. so I just found out that getting a logs, results in the hosted pipeline is not yet implemented so we don't have URL yet. I'll add it in AC for getting pipeline artifacts.

Comment (amisstea) [reply ansible/roles/operator-pipeline/templates/openshift/tasks/upload-artifact.yml:78]: Ok, thanks.

**Review by amisstea — APPROVED** [review]

**Pull request merged**